### ISSUE TITLE
[HOLD for payment 2024-08-02] [HOLD for payment 2024-08-01] [$250] MEDIUM: [Debugability] Add profile trace download option to Desktop and Web

### ISSUE THREAD

**Issue by muttmuure** [Expensify/App#43363]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?
e.g. memory consumption, storage read/write times, React native bridge concerns, inefficient React component rendering, etc.

Desktop and web performance is currently difficult to evaluate compared with mobile because we don't support profile tracing 

## What is the impact of this on end-users?
List specific user experiences that will be improved by solving this problem e.g. app boot time, time to for some interaction to complete, etc.

Desktop and web users are unable to have their performance issues fully triaged. 

## List any benchmarks that show the severity of the issue
Please also provide exact steps taken to collect metrics above if any so we can independently verify the results.

Triaging [this issue](https://github.com/Expensify/App/issues/42558) would benefit from having a profile trace

## Proposed solution (if any)
Please list out the steps you think we should take to solve this issue.

Add web support to react-native-release-profiler
Make the "share profile" flow work on web as well

Add profile trace functionality to the desktop and web apps, in the Troubleshoot menu and four-finger-tap menu

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)
Note: These should be the same as the benchmarks collected before any changes.

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**
**Issue reported by:** @muttmuure 
**Slack conversation:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018b1cc2a053306cc3</li>
        <li>Upwork Job ID: 1801178849066862909</li>
        <li>Last Price Increase: 2024-06-20</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @jayeshmangwani</details>

**Comment by jayeshmangwani** [comment]
Waiting on proposal, bumped on [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q/p1718738479785389)

**Comment by hannojg** [comment]
Hey, Hanno from Margelo here 👋 we build react-native-release-profiler the library we use on mobile for recording performance profiles. We can enable recording profiles for desktop (and web), by adding web support to the library.
Feel free to assign me!

**Comment by kirillzyusko** [comment]
Hey, Kiryl from Margelo here 👋 

@hannojg asked me to work on this task. Can you please assign me on this task? 👀

**Comment by kirillzyusko** [comment]
I'm working on this task. At the moment I can capture profile traces and I'm discovering the best option for representing this data (obviously it has to match Google Chrome parser standards). Right now I have two options:
- to use a custom transformer;
- to convert to Hermes profiler format and then use hermes transformer.

I think tomorrow or the day after tomorrow the PR will be ready for review 👀

**Comment by jayeshmangwani** [comment]
Not Overdue! Kirill is working on this issue. The last update is [here](https://github.com/Expensify/App/issues/43363#issuecomment-2187203854), and the PR will be ready for review in a day or two.

**Comment by kirillzyusko** [comment]
Still working on it! Basic functionality works - I just need to verify, that source map support works as well and it can correctly show function names for minified code 👀

**Comment by jayeshmangwani** [comment]
PR is in [Draft](https://github.com/Expensify/App/pull/44432)

**Comment by muttmuure** [comment]
How is this coming along? Do you need an internal reviewer to help out too?

**Comment by hannojg** [comment]
Hey, we are done with:
- Adding web support to react-native-release-profiler
- Updating and adapting to the new version in expensify

The only thing we are currently working on is getting the source map symbolication to work correctly.
On web/desktop, when we make a release build the js code gets obfuscated / the names get mangled. We are currently trying to see if we can create a source map with de-mangled names. Working on this today with Kiryl, should have a new update by EOD!

**Comment by muttmuure** [comment]
amazing

**Comment by hannojg** [comment]
Hey, sorry our availability was a bit limited today - no new updates yet. I'll look into this first thing tomorrow morning!

**Comment by hannojg** [comment]
Alright, so here are our findings:

- On web the code gets minified, which means that the function names will be mangled (e.g. `toggleCreateButton` becomes `dK`)
- We have a source map for the minified function names
- The cpu profile recorded will record the production code with the minified function names
- There is no easy way to convert the function names for the final profile back to their regular names

That being said, the function names might look weird, but we still get the file name + line, which is pretty usable:

<img width="492" alt="image" src="https://github.com/Expensify/App/assets/16821682/37783f59-39d9-4640-9c42-a8df16609b31">

Note: We can build a script that will use [some open source solutions](https://github.com/getsentry/symbolic/tree/master/symbolic-sourcemapcache) to bring back the function names, but that would probably be a endeavour of multiple days and we didn't want to block this issue by this (as there is no direct tool for this and we'd need to build one). We can maybe open a new issue for that (would take 1-2 days to get that working)?

**Next steps on the PR are:**

- Upload source maps for web in the GitHub actions
- Cleanup our PR
- Merge :shipit:

**Comment by kirillzyusko** [comment]
@muttmuure this PR https://github.com/Expensify/App/pull/44432 is ready for review 👀

**Comment by lschurr** [comment]
Payment summary is correct. No payments in Upwork. Closing.

**Comment by jayeshmangwani** [comment]
@lschurr Issue is not ready to be paid out yet. The original PR was reverted due to a regression, and a new PR was just merged 3 days ago. Please keep this issue open until the new PR has been in production for 7 days.

**Comment by lschurr** [comment]
Can you link the new PR here @jayeshmangwani?

**Comment by kirillzyusko** [comment]
@lschurr it's a link for new PR https://github.com/Expensify/App/pull/46133 🙌

**Comment by muttmuure** [comment]
https://github.com/Expensify/App/issues/46929

**Comment by lschurr** [comment]
Payment summary:

Due to the regression, payments are half.

- Contributor: @kirillzyusko (no payment needed)
- Reviewer: @jayeshmangwani $125 via NewDot

**Comment by jayeshmangwani** [comment]
Requested on ND as per https://github.com/Expensify/App/issues/43363#issuecomment-2289077081

**Comment by JmillsExpensify** [comment]
$125 approved for @jayeshmangwani
